pkg/controller/seed-controller-manager/initial-application-installation-controller/controller.go

```diff
@@ -0,0 +1,231 @@
+	"k8c.io/kubermatic/v2/pkg/provider"
+	"k8c.io/kubermatic/v2/pkg/version/kubermatic"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/record"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	ControllerName = "kkp-initial-application-installation-controller"
+)
+
+// UserClusterClientProvider provides functionality to get a user cluster client.
+type UserClusterClientProvider interface {
+	GetClient(ctx context.Context, c *kubermaticv1.Cluster, options ...clusterclient.ConfigOption) (ctrlruntimeclient.Client, error)
+}
+
+type Reconciler struct {
+	ctrlruntimeclient.Client
+
+	workerName                    string
+	recorder                      record.EventRecorder
+	seedGetter                    provider.SeedGetter
+	userClusterConnectionProvider UserClusterClientProvider
+	log                           *zap.SugaredLogger
+	versions                      kubermatic.Versions
+}
+
+func Add(ctx context.Context, mgr manager.Manager, numWorkers int, workerName string, seedGetter provider.SeedGetter, userClusterConnectionProvider UserClusterClientProvider, log *zap.SugaredLogger, versions kubermatic.Versions) error {
+	reconciler := &Reconciler{
+		Client: mgr.GetClient(),
+
+		workerName:                    workerName,
+		recorder:                      mgr.GetEventRecorderFor(ControllerName),
+		seedGetter:                    seedGetter,
+		userClusterConnectionProvider: userClusterConnectionProvider,
+		log:                           log,
+		versions:                      versions,
+	}
+
+	c, err := controller.New(ControllerName, mgr, controller.Options{
+		Reconciler:              reconciler,
+		MaxConcurrentReconciles: numWorkers,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create controller: %w", err)
+	}
+
+	predicate := predicateutil.MultiFactory(predicateutil.TrueFilter, func(o ctrlruntimeclient.Object) bool {
+		_, exists := o.GetAnnotations()[v1.InitialApplicationInstallationsRequestAnnotation]
+		return exists
+	}, predicateutil.TrueFilter)
+
+	if err := c.Watch(&source.Kind{Type: &kubermaticv1.Cluster{}}, &handler.EnqueueRequestForObject{}, predicate); err != nil {
+		return fmt.Errorf("failed to create watch: %w", err)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	cluster := &kubermaticv1.Cluster{}
+	if err := r.Get(ctx, request.NamespacedName, cluster); err != nil {
+		if kerrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	if cluster.DeletionTimestamp != nil {
+		// Cluster is queued for deletion; no action required
```

Comment:
I think we should log a line here to indicate to users that application reconciliation is stopped, because the cluster is marked for deletion